### ISSUE TITLE
github upload-results fails when uploading large SARIF with incorrect error message

### ISSUE THREAD

**Issue by fergusodonnell** [github/codeql#15209]
github upload-results fails when when the SARIF file is 28MB. 
The upload-results command appears to succeed but the failure  is evident upon checking the results of the upload with:
https://api.github.com/repos/Sage-AutoEntry/OCRex.AutoEntry/code-scanning/sarifs/19f56d2a-9dd5-11ee-99f0-07a8c738d60c

The following is the response:
  GitHub cannot process SARIF file. Status: Failed. Errors: SARIF URI scheme "d" did not match the checkout URI scheme "file"

**Comment by jketema** [comment]
Hi @fergusodonnell,

We'll need a bit more context to help you. Could you tell us where the SARIF file is coming from?

**Comment by fergusodonnell** [comment]
Hi @jketema, thanks for responding.  
The SARIF file was generated from the analysis of a DB produced using 

Using a command such as the following:
codeql.exe database analyze C:\\Work\\Repos\\<REPO NAME>\.codeQlOutput\database\csharp --format=sarif-latest --sarif-category C:\\Work\\Repos\\<REPO NAME>/src/_csharp --output C:\\Work\\Repos\\<REPO NAME>\.codeQlOutput\analysisResults\resultscsharp.json --threads 2 --ram 4000

The first step was to generate the database using a command similar to the following.

A work-around was to reduce the query packs referenced in the CodeQlConfig.yml file but this is not satisfactory when we need "Security and Quality".
codeql.exe database create C:\\Work\\Repos\\<REPO NAME>\.codeQlOutput\database\csharp --language csharp --source-root C:\\Work\\Repos\\<REPO NAME>\\src --overwrite --threads 1 --ram 4000 --codescanning-config C:\work\repos\<REPO NAME>\CodeQlConfig.yml

**Comment by jketema** [comment]
What happens when you pass the `[--max-paths=](https://docs.github.com/en/code-security/codeql-cli/codeql-cli-manual/database-analyze#--max-pathsmaxpaths)` option to `database analyze` and set it to 0? I wonder if this is a size related problem or not, and setting 0 should produce a significantly smaller SARIF file. I'm just trying to diagnose here; not suggesting setting 0 as a permanent solution.

**Comment by fergusodonnell** [comment]
I will try this, thanks. However the response from the operations do not indicate that a hard or soft limit is exceeded.

**Comment by aibaars** [comment]
The error about the scheme 'd' sounds like the problem with a URL starting with D:/ . Possibly a Windows file path without a file:/ prefix

**Comment by fergusodonnell** [comment]
> The error about the scheme 'd' sounds like the problem with a URL starting with D:/ . Possibly a Windows file path without a file:/ prefix

Thanks @aibaars I will check on this too: 

Within the SARIF file there appears to be a mixture of URI styles depending on the rule being evaluated. : 
For the Generic Catch clause it has D: directly whereas for "setter call argument" it has file://

@aibaars 

             "physicalLocation": {
                "artifactLocation": {
                  "uri": "file:/D:/Windows_General_Build_2/temp/....<filename>.dll",
                  
          "physicalLocation": {
                "artifactLocation": {
                  "uri": "D:\\Windows_General_Build_2\\...<Filename>.cs",

**Comment by aibaars** [comment]
Thanks! I think this one is the problem `"uri": "D:\\Windows_General_Build_2\\...<Filename>.cs",` . Could you sent the entire JSON "result" object containing that location for more context?

**Comment by fergusodonnell** [comment]
> Thanks! I think this one is the problem `"uri": "D:\\Windows_General_Build_2\\...<Filename>.cs",` . Could you sent the entire JSON "result" object containing that location for more context?

Here is a sample, slightly redacted. Is this enough ? 

{
  "ruleId": "cs/catch-of-all-exceptions",
  "ruleIndex": 97,
  "rule": {
    "id": "cs/catch-of-all-exceptions",
    "index": 97
  },
  "message": {
    "text": "Generic catch clause."
  },
  "locations": [
    {
      "physicalLocation": {
        "artifactLocation": {
          "uri": "D:\\Windows_General_Build_2\\work\\4d72a66d0adeb31b/src//MyLib/MyFile.cs",
          "uriBaseId": "%SRCROOT%",
          "index": 1168
        },
        "region": {
          "startLine": 122,
          "startColumn": 17,
          "endLine": 125,
          "endColumn": 18
        }
      }
    }
  ],
  "partialFingerprints": {
    "primaryLocationLineHash": "5dce282f95e85853:1",
    "primaryLocationStartColumnFingerprint": "0"
  }
}

**Comment by aibaars** [comment]
> Here is a sample, slightly redacted. Is this enough ?

Yes, I think so. That URL looks pretty weird, a mix of `\` and `/` and even a `//` . Something seems to have gone terribly wrong with URL/path normalization. Thanks for reporting!

**Comment by fergusodonnell** [comment]
Thanks @aibaars ,
I think that was the issue. Having altered all the paths in the very large SARIF files so they contain / rather than \\ and "file:///D: rather than "D:  the upload works correctly.  Not sure if the issue is because of the source files but a step to bulk replace the paths should fix it in our automation. 
Ideally CodeQL would perform consistently however in terms of ensuring that the SARIF file generated has consistent paths, so I'll leave the issue open.

**Comment by aibaars** [comment]
@fergusodonnell That's great! I think CodeQL should have dropped those URLs because they are outside the `--source-root` . GitHub's Code Scanning UI can't link those URLs to source files in the repository. 

If those files can actually be found in the repository, then you could try making the URLs relative, something like `"uri": "src/MyLib/MyFile.cs"` or `"uri": "MyLib/MyFile.cs"` depending on the folder structure of the repository.

**Comment by fergusodonnell** [comment]
@aibaars: To conclude: 
All files were within the folder passed in --source-root. However the absolute path passed in --source-root, while at the time necessary for the build seemed to cause the absolute paths to be created in the DB and SARIF file. 
Changing the --source-root to a folder within the checkout path allows relative paths to be generated thereby solving the issue. 

Thanks again to you and @jketema for the help.

**Comment by fergusodonnell** [comment]
@aibaars: To conclude:
All files were within the folder passed in --source-root. However the absolute path passed in --source-root, while at the time necessary for the build seemed to cause the absolute paths to be created in the DB and SARIF file.
Changing the --source-root to a folder within the checkout path allows relative paths to be generated thereby solving the issue.

Thanks again to you and @jketema for the help.
